keps/sig-cloud-provider/1959-service-lb-class-annotation/README.md

```diff
@@ -0,0 +1,203 @@
+
+[kubernetes.io]: https://kubernetes.io/
+[kubernetes/enhancements]: https://git.k8s.io/enhancements
+[kubernetes/kubernetes]: https://git.k8s.io/kubernetes
+[kubernetes/website]: https://git.k8s.io/website
+
+## Summary
+
+When Service Type=LoadBalancer is enabled by a Kubernetes cloud provider, it is a global
+configuration that applies for all Service Type=LoadBalancer resources in a given cluster.
+This becomes problematic if users want to leverage multiple Service Type=LoadBalancer
+implementations in a given cluster.
+
+The new [Services APIs](https://github.com/kubernetes-sigs/service-apis) addresses this already
+with the GatewayClass resource. However, until Gateway/GatewayClass APIs become mature, we should
+support similar functionality for Services of Type=LoadBalancer. Introducing a new resource like
+`ServiceClass` is probably not worthwhile given that there are new APIs already in development.
+This KEP proposes a light-weight approach for Service Type=LoadBalancer by introducing a Service
+annotation `service.kubernetes.io/load-balancer-class`.
+
+## Motivation
+
+The main use-case for this feature is being able to support multiple Service Type=LoadBalancer
+implementations in a cluster, as different workloads may want to leverage different loadbalancer
+providers based on efficiency, availability, cost and other factors.
+
+For example, a cluster admin may want to use a public load balancer from a cloud provider
+for workloads that must be assigned a publically routable address, but they may want to
+enable a lower-cost solution for workloads that are only internally accessible.
+
+### Goals
+
+* allow users to opt-out of the Service Type=LoadBalancer implementation by the cloud provider.
+* allow multiple implementations of Service Type=LoadBalancer in a given cluster.
+
+### Non-Goals
+
+* performance improvements for Service Type=LoadBalancer.
+* changing any other existing behaviors for Service Type=LoadBalancer aside from being able
+to disabling it from the cloud provider.
+
+## Proposal
+
+This KEP proposes to add a new Service annotation `service.kubernetes.io/load-balancer-class`
+that allows for multiple implementations of Service Type=LoadBalancer in a cluster.
+
+### User Stories (Optional)
+
+#### Story 1
+
+As a cluster admin:
+* I want to use my cloud provider's public load balancer service for applications that require
+public ingress.
+* I want to use my own load balancing solution for any applications that only talk internally
+within my own network because I want to save costs.
+
+#### Story 2
+
+As an application developer:
+* I MUST use a hardware-based loadbalancer for certain applications due to specific protocols
+only available there.
+* I want to use the cloud provider's default load balancer for any applications that do not
+rely on protocols from hardware load balancers.
```

Comment:
Is this really a use case? Public clouds generally wouldn't let you have custom network hardware, and private clouds generally don't implement their own load balancers...

Reply:
> private clouds generally don't implement their own load balancers...

This isn't true, the OpenStack cloud provider for example has their own load balancer implementations and newer vSphere versions come with Service Type=LoadBalancer using platforms like NSX.

Agreed that the use-case is not as common in public clouds though, but load balancing vendors like F5 do support running on public clouds so it's not unheard of.